lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp

```diff
@@ -785,6 +890,18 @@ void swift::performSILDiagnoseUnreachable(SILModule *M, SILModuleTransform *T) {
     // Remove unreachable blocks.
     removeUnreachableBlocks(Fn, *M, &State);
 
+    // Gather exit blocks to check for naive self-recursion.
+    llvm::SmallPtrSet<SILBasicBlock*, 4> ExitBlocks;
+    for (auto &BB : Fn) {
+      if (BB.getTerminator()->isFunctionExiting())
+        ExitBlocks.insert(&BB);
+    }
+
+    // Diagnose infinitely recursive applies.
+    if (checkInfinitelyRecursiveApplies(Fn, ExitBlocks))
+      diagnose(Fn.getModule().getASTContext(),
```

Comment:
I wonder if it would make sense to extend such an analysis to detect even a more general form of recursion, i.e. not only the self-recursion? It may turn out that it is just too expensive. But may be it is not so expensive. 

You would  essentially need to build a call-graph, I guess. We have a `BasicCalleeAnalysis` that could help here.

WDYT?

Reply:
It would be nice ™ to look into mutual recursion of two functions because of [this SR](https://bugs.swift.org/browse/SR-5224).

Reply:
@CodaFi @swiftix I do not think that we build the call graph at -Onone ever (maybe my memory is wrong). If we are to do so, please check compile time (if it becomes an issue, I imagine you can basically add a limit on the size of the call graph being exposed perhaps).